### PR TITLE
Feature/null analysis external annotations

### PR DESCRIPTION
**Pull request by tboue**
split from https://github.com/MicroEJ/docs/pull/329 .
Part on java 8 won't be added because developers work with java 7.
Part on external-annotations is kept and moved to null annotation page.

### CODE REVIEW

File: ApplicationDeveloperGuide/nullAnalysis.rst [hunk 1]
```diff
@@ -76,7 +76,7 @@ To enable the Null Analysis tool, a :ref:`module project <mmm_module_skeleton>`
   This option allows to fully ignore Null Analysis errors in advanced cases using ``@SuppressWarnings("null")`` annotation.
 
 
-If you have multiple projects to configure, you can then copy the content of the ``.settings`` folder to an other :ref:`module project <mmm_module_skeleton>`.
+If you have multiple projects to configure, you can then copy the content of the ``.settings`` folder to another :ref:`module project <mmm_module_skeleton>`.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nullAnalysis.rst:79]: Replace `you` formulations with a passive sentence.

File: ApplicationDeveloperGuide/nullAnalysis.rst [hunk 2]
```diff
@@ -103,6 +103,124 @@ Here are some examples to illustrate the philosophy:
 Implementations are left unchanged and still comply with the Javadoc description whether the Null Analysis is enabled or not. 
 So if these additional constraints are not acceptable for your project, please disable Null Analysis.
 
+External null annotations
+-------------------------
+
+In some situations, the null analysis will miss potential null access while using unannotated external libraries.
+For instance : an ArrayList can contain null values, hence the `get(int)` method can return null values. However, this won't be taken into account by default during null analysis.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nullAnalysis.rst:110]:
```suggestion
For instance : an ``ArrayList`` can contain null values, hence the `get(int)` method can return null values. However, this won't be taken into account by default during the null analysis.
```

File: ApplicationDeveloperGuide/nullAnalysis.rst [hunk 3]
```diff
@@ -103,6 +103,124 @@ Here are some examples to illustrate the philosophy:
...
+
+.. figure:: images/null_analysis_ArrayList_annotation_before.png
+
+Manually annotate the `get(int)` method return type with @Nullable allows to detect the potential null access:
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nullAnalysis.rst:114]: rst uses double backquotes for the code not one (markdown)

File: ApplicationDeveloperGuide/nullAnalysis.rst [hunk 4]
```diff
@@ -103,6 +103,124 @@ Here are some examples to illustrate the philosophy:
...
+  - **List<T> -> Ljava/util/List<TT;> -> L0java/util/List<T0T;>**
+  - (ArrayList's to Array method) **public <T> T[] toArray(T[] a) -> <T:Ljava/lang/Object;>([TT;)[TT; -> <0T:Ljava/lang/Object;>([0T0T;)[0T0T;**
+
+Update changes
+~~~~~~~~~~~~~~
+
+When editing ``.eea`` files manually it may be needed to clean the project **Project->Clean** for the changes to take place.
+If the changes still aren't taken into account, the IDE must be restarted.
+
+It's also worth the try with automatic annotations if any problem is encountered.
```
Comment (gjadi-mej) [inline ApplicationDeveloperGuide/nullAnalysis.rst:223]: What do you mean by this?

**Review by gjadi-mej — CHANGES_REQUESTED** [review]
first round of review, I'll come back again later.

File: ApplicationDeveloperGuide/nullAnalysis.rst (hunk 1 above)
Comment (tboue) [reply ApplicationDeveloperGuide/nullAnalysis.rst:79]: This part is constructed like a tutorial won't the `you` formulation be ok ?

**Comment by tboue** [comment]
On hold : external annotations don't seem possible by default in MicroEJ's SDK with ej.annotation package

**Comment by tboue** [comment]
Everything moved to private documentation because null annotations are already performed on MicroEJ's public libraries.

**Pull request closed without merging**